File: app/src/main/java/com/tomclaw/appsend/util/MultipartStream.java
package com.tomclaw.appsend.util;

import android.util.Log;

import java.io.IOException;
import java.io.InputStream;
import java.io.InterruptedIOException;
import java.io.OutputStream;

/**
 * Created by solkin on 13.01.15.
 */
public class MultipartStream extends OutputStream {

    private static final String TAG = "multipart";
    private static final int BUFFER_SIZE = 128 * 1024;

    private final String boundary;
    private final OutputStream outputStream;

    private boolean isSetLast = false;
    private boolean isSetFirst = false;

    public MultipartStream(OutputStream outputStream, String boundary) {
        this.outputStream = outputStream;
        this.boundary = boundary;
    }

    public void writeFirstBoundaryIfNeeds() {
        if (!isSetFirst) {
            try {
                outputStream.write(("--" + boundary + "\r\n").getBytes());
            } catch (final IOException e) {
                Log.e(TAG, e.getMessage(), e);
            }
        }
        isSetFirst = true;
    }

    public void writeLastBoundaryIfNeeds() {
        if (isSetLast) {
            return;
        }
        try {
            outputStream.write(("\r\n--" + boundary + "--\r\n").getBytes());
        } catch (final IOException e) {
            Log.e(TAG, e.getMessage(), e);
        }
        isSetLast = true;
    }

    public void writePart(final String key, final String value) {
        writeFirstBoundaryIfNeeds();
        try {
            outputStream.write(("Content-Disposition: form-data; name=\"" + key + "\"\r\n").getBytes());
            outputStream.write("Content-Type: text/plain; charset=UTF-8\r\n".getBytes());
            outputStream.write("Content-Transfer-Encoding: 8bit\r\n\r\n".getBytes());
            outputStream.write(value.getBytes());
            outputStream.write(("\r\n--" + boundary + "\r\n").getBytes());
        } catch (final IOException e) {
            Log.e(TAG, e.getMessage(), e);
        }
    }

    public void writePart(final String key, final String fileName, final InputStream inputStream,
                          String type, ProgressHandler callback) {
        writeFirstBoundaryIfNeeds();
        try {
            type = "Content-Type: " + type + "\r\n";
            outputStream.write(("Content-Disposition: form-data; name=\"" + key + "\"; filename=\"" + fileName + "\"\r\n").getBytes());
            outputStream.write(type.getBytes());
            outputStream.write("Content-Transfer-Encoding: binary\r\n\r\n".getBytes());

            int cache;
            long sent = 0;
            byte[] buffer = new byte[BUFFER_SIZE];
            while ((cache = inputStream.read(buffer, 0, BUFFER_SIZE)) != -1) {
                outputStream.write(buffer, 0, cache);
                outputStream.flush();
                sent += cache;
                callback.onProgress(sent);
                if (Thread.interrupted()) {
                    throw new InterruptedException();
                }
            }
            outputStream.write(("\r\n--" + boundary + "\r\n").getBytes());
        } catch (InterruptedIOException ex) {
            Log.e(TAG, "[upload] IO interruption while application downloading", ex);
            callback.onCancelled(ex);
        } catch (InterruptedException ex) {
            Log.e(TAG, "[upload] Interruption while application downloading", ex);
            callback.onCancelled(ex);
        } catch (final Throwable e) {
            Log.e(TAG, e.getMessage(), e);
            callback.onError(e);
        } finally {
            try {
                inputStream.close();
            } catch (final IOException e) {
                Log.e(TAG, e.getMessage(), e);
            }
        }
    }

    @Override
    public void write(int oneByte) throws IOException {
        outputStream.write(oneByte);
    }

    @Override
    public void flush() throws IOException {
        outputStream.flush();
    }

    public interface ProgressHandler {
        void onProgress(long sent);

        void onError(Throwable e);

        void onCancelled(Throwable e);
    }
}

File: app/src/main/java/com/tomclaw/appsend/util/MultipartStream.kt
package com.tomclaw.appsend.util

import android.util.Log
import java.io.IOException
import java.io.InputStream
import java.io.InterruptedIOException
import java.io.OutputStream

/**
 * Created by solkin on 13.01.15.
 */
class MultipartStream(
    private val outputStream: OutputStream,
    private val boundary: String
) : OutputStream() {

    private var isSetLast = false
    private var isSetFirst = false

    fun writeFirstBoundaryIfNeeds() {
        if (!isSetFirst) {
            try {
                outputStream.write("--$boundary\r\n".toByteArray())
            } catch (e: IOException) {
                Log.e(TAG, e.message, e)
            }
        }
        isSetFirst = true
    }

    fun writeLastBoundaryIfNeeds() {
        if (isSetLast) {
            return
        }
        try {
            outputStream.write("\r\n--$boundary--\r\n".toByteArray())
        } catch (e: IOException) {
            Log.e(TAG, e.message, e)
        }
        isSetLast = true
    }

    fun writePart(key: String, value: String) {
        writeFirstBoundaryIfNeeds()
        try {
            outputStream.write("Content-Disposition: form-data; name=\"$key\"\r\n".toByteArray())
            outputStream.write("Content-Type: text/plain; charset=UTF-8\r\n".toByteArray())
            outputStream.write("Content-Transfer-Encoding: 8bit\r\n\r\n".toByteArray())
            outputStream.write(value.toByteArray())
            outputStream.write("\r\n--$boundary\r\n".toByteArray())
        } catch (e: IOException) {
            Log.e(TAG, e.message, e)
        }
    }

    fun writePart(
        key: String,
        fileName: String,
        inputStream: InputStream,
        type: String,
        callback: ProgressHandler
    ) {
        writePart(key, fileName, writer = { outputStream ->
            inputStream.use { stream ->
                var cache: Int
                var sent: Long = 0
                val buffer = ByteArray(BUFFER_SIZE)
                while (stream.read(buffer, 0, BUFFER_SIZE).also { cache = it } != -1) {
                    outputStream.write(buffer, 0, cache)
                    outputStream.flush()
                    sent += cache.toLong()
                    callback.onProgress(sent)
                    if (Thread.interrupted()) {
                        throw InterruptedException()
                    }
                }
            }
        }, type, callback)
    }

    fun writePart(
        key: String,
        fileName: String,
        writer: (OutputStream) -> (Unit),
        type: String,
        callback: ProgressHandler
    ) {
        var type = type
        writeFirstBoundaryIfNeeds()
        try {
            type = "Content-Type: $type\r\n"
            outputStream.write("Content-Disposition: form-data; name=\"$key\"; filename=\"$fileName\"\r\n".toByteArray())
            outputStream.write(type.toByteArray())
            outputStream.write("Content-Transfer-Encoding: binary\r\n\r\n".toByteArray())
            writer.invoke(outputStream)
            outputStream.write("\r\n--$boundary\r\n".toByteArray())
        } catch (ex: InterruptedIOException) {
            Log.e(TAG, "[upload] IO interruption while application downloading", ex)
            callback.onCancelled(ex)
        } catch (ex: InterruptedException) {
            Log.e(TAG, "[upload] Interruption while application downloading", ex)
            callback.onCancelled(ex)
        } catch (e: Throwable) {
            Log.e(TAG, e.message, e)
            callback.onError(e)
        }
    }

    @Throws(IOException::class)
    override fun write(oneByte: Int) {
        outputStream.write(oneByte)
    }

    @Throws(IOException::class)
    override fun flush() {
        outputStream.flush()
    }

    interface ProgressHandler {
        fun onProgress(sent: Long)
        fun onError(ex: Throwable)
        fun onCancelled(ex: Throwable)
    }

}

private const val TAG = "multipart"
private const val BUFFER_SIZE = 128 * 1024
